### PR TITLE
[DLP] Implemented dlp_k_anonymity_with_entity_id

### PR DESCRIPTION
**Pull request by Avani-Thakker-Crest**
## Description
Implemented dlp_k_anonymity_with_entity_id. Json equivalent: https://cloud.google.com/dlp/docs/visualizing_re-id_risk#step_1_calculate_k-anonymity_on_the_dataset

Fixes #<ISSUE-NUMBER>

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [x] **Tests** pass:   `nox -s py-3.9` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [x] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: dlp/snippets/risk.py [hunk 1]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
 # [END dlp_k_anonymity]
 
 
+# [START dlp_k_anonymity_with_entity_id]
+import time  # noqa: I100,  F811, E402
+from typing import List  # noqa: E402, F811
+
+import google.cloud.dlp  # noqa: I100, F811, E402
```
Comment (rsamborski) [inline dlp/snippets/risk.py:390]: issue: is this a correct import? You use `google.cloud.dlp_v2` in line 423

File: dlp/snippets/risk.py [hunk 2]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+    parent = f"projects/{project}/locations/global"
+
+    # Call API to start risk analysis job.
+    response = dlp.create_dlp_job(
+        request={
+            "parent": parent,
+            "risk_job": risk_job,
+        }
+    )
+    print("Inspection Job started : {}".format(response.name))
```
Comment (rsamborski) [inline dlp/snippets/risk.py:485]: nit: please use an f-string

File: dlp/snippets/risk.py [hunk 3]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+
+    job_name = response.name
+
+    # Waiting for a maximum of 15 minutes for the job to be completed.
+    job = dlp.get_dlp_job(request={"name": job_name})
+    no_of_attempts = 30
+    while no_of_attempts > 0:
```
Comment (rsamborski) [inline dlp/snippets/risk.py:492]: question: is this the recommended way to wait for DLP's job results? It seems quite cumbersome to poll the result every 30 seconds. Does the library support long-running operations in similar way as compute does ([see here](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/compute/client_library/snippets/operations/wait_for_extended_operation.py) for an example of handling an LRO).

File: dlp/snippets/risk.py [hunk 4]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+    print(f"Job name: {job.name}")
+    histogram_buckets = (
+        job.risk_details.k_anonymity_result.equivalence_class_histogram_buckets
+    )
+    # Print bucket stats
+    for i, bucket in enumerate(histogram_buckets):
+        print(f"Bucket {i}:")
+        if bucket.equivalence_class_size_lower_bound:
+            print(
+                "   Bucket size range: [{}, {}]".format(
```
Comment (rsamborski) [inline dlp/snippets/risk.py:525]: suggestion: please use f-strings

File: dlp/snippets/risk.py [hunk 5]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+                    bucket.equivalence_class_size_lower_bound,
+                    bucket.equivalence_class_size_upper_bound,
+                )
+            )
+            for value_bucket in bucket.bucket_values:
+                print(
+                    "   Quasi-ID values: {}".format(
```
Comment (rsamborski) [inline dlp/snippets/risk.py:532]: (ditto) suggestion: please use f-strings

File: dlp/snippets/risk.py [hunk 6]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+                        get_values(value_bucket.quasi_ids_values[0])
+                    )
+                )
+                print(
+                    "   Class size: {}".format(
```
Comment (rsamborski) [inline dlp/snippets/risk.py:537]: (ditto) suggestion: please use f-strings

File: dlp/snippets/risk.py [hunk 7]
```diff
@@ -782,6 +943,49 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+        "source_dataset_id",
+        help="The id of the dataset to inspect.",
+    )
+    k_anonymity_entity_parser.add_argument(
+        "source_table_id",
+        help="The id of the table to inspect.",
+    )
+    k_anonymity_entity_parser.add_argument(
+        "entity_id",
+        help="It enables you to accurately determine k-anonymity.",
```
Comment (rsamborski) [inline dlp/snippets/risk.py:968]: suggestion: (ditto) please update when you update the comment per suggestion above.

**Review by rsamborski — CHANGES_REQUESTED** [review]
Reviewed from sample quality perspective.

File: dlp/snippets/risk.py (hunk 1 above)
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py:390]: that's how its used in other samples, yet I have updated the same

File: dlp/snippets/risk.py (hunk 2 above)
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py]: done

File: dlp/snippets/risk.py (hunk 4 above)
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py]: done

File: dlp/snippets/risk.py (hunk 5 above)
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py]: done

File: dlp/snippets/risk.py (hunk 6 above)
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py]: done

File: dlp/snippets/risk.py [hunk 8]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
 # [END dlp_k_anonymity]
 
 
+# [START dlp_k_anonymity_with_entity_id]
+import time  # noqa: I100,  F811, E402
+from typing import List  # noqa: E402, F811
+
+import google.cloud.dlp  # noqa: I100, F811, E402
+import google.cloud.pubsub  # noqa: I100, F811, E402
```
[the remark this answers is not in the document]
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py]: thank you for pointing, updated.

File: dlp/snippets/risk.py (hunk 3 above)
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py:490]: Yes, I also feel this looks quite cumbersome. Here, we are calling get_dlp_job, which will give us the response at that instant, does it account to a long-running jobs? create_dlp_job or get_dlp_job, both of them doesn't seem to be long running operation. I couldn't find any efficient way to poll the request till we get the status - "Done". Though I am still looking at it and will update the code accordingly if I find any good way to handle this. Thank you!

File: dlp/snippets/risk.py [hunk 9]
```diff
@@ -383,6 +383,167 @@ def callback(message: google.cloud.pubsub_v1.subscriber.message.Message) -> None
...
+
+    """Uses the Data Loss Prevention API to compute the k-anonymity using entity_id
+        of a column set in a Google BigQuery table.
+    Args:
+        project: The Google Cloud project id to use as a parent resource.
+        source_table_project_id: The Google Cloud project id where the BigQuery table
+            is stored.
+        source_dataset_id: The id of the dataset to inspect.
+        source_table_id: The id of the table to inspect.
+        entity_id: It enables you to accurately determine k-anonymity.
```
[the remark this answers is not in the document]
Comment (rsamborski) [reply dlp/snippets/risk.py:413]: It's rather a linguistic suggestion. Other arguments have a description explaining what should be passed, i.e. "The id of the dataset to inspect.", "The id of the table to inspect.". While entity_id states "It enables you to accurately determine k-anonymity.". As a person who never used this API it's unclear what I should pass in the entity_id, should it be some id of an existing entity? Is it a hash used for anonymity?
Does it make sense?

[the remark this answers is not in the document]
Comment (Avani-Thakker-Crest) [reply dlp/snippets/risk.py]: Right got it, Thank you for detailed description :) I have updated accordingly, please check.

**Review by rsamborski — APPROVED** [review]
LGTM. It would be great if you figure out a better solution for polling the job, but if it's not possible then I am ok to merge as is.

File: dlp/snippets/risk.py (hunk 3 above)
Comment (annadata) [reply dlp/snippets/risk.py:492]: @realjordanna Do you have any recommendations? Are there any convenience methods in the SDK that help. This has come up in the past as well.

**Review by annadata — APPROVED** [review]

**Pull request merged**